### PR TITLE
expression: fix incorrect result of logical operators when isTrue/ isFalse function is pushed down

### PR DESCRIPTION
**Pull request by Reminiscent**
### What problem does this PR solve?
Close: #15718  #15987 
Related PR: #12173 
Need to review [TiPB176](https://github.com/pingcap/tipb/pull/176) first
Related to [#13962](https://github.com/pingcap/tidb/pull/13962)

### What is changed and how it works?
See an example of this [comment](https://github.com/pingcap/tidb/issues/15718#issuecomment-605763963)
Before this PR:
```
MySQL [test]> select a and b as c, a, b from tt order by c limit 4;
+---+------+------+
| c | a    | b    |
+---+------+------+
| NULL |    7 | NULL |
| NULL | NULL | 2    |
| 0 |    0 | 2    |
| 0 | NULL | w    |
+---+------+------+
4 rows in set, 4 warnings (0.00 sec)

MySQL [test]> select a and b as c, a, b from tt order by c;
+---+------+------+
| c | a    | b    |
+---+------+------+
| NULL | NULL | NULL |
| NULL |    7 | NULL |
| NULL | NULL | 2    |
| NULL | NULL | 2    |
| 0 | NULL | w    |
| 0 |    0 | 2    |
| 0 |    7 | f    |
| 1 |    7 | 3    |
+---+------+------+
8 rows in set, 2 warnings (0.01 sec)
```

After this PR:
```
mysql> create table tt(a decimal(10, 0), b varchar(1));
Query OK, 0 rows affected (0.02 sec)

mysql> insert into tt values(0, '2'), (7, null), (7, '3'), (NULL, 'w'), (NULL, '2'), (NULL, '2'), (NULL, NULL), (7, 'f');
Query OK, 8 rows affected (0.00 sec)
Records: 8  Duplicates: 0  Warnings: 0

mysql> select a and b as c, a, b from tt order by c limit 4;
+---+------+------+
| c | a    | b    |
+---+------+------+
| NULL | NULL | NULL |
| NULL | NULL | 2    |
| NULL | NULL | 2    |
| NULL |    7 | NULL |
+---+------+------+
4 rows in set, 2 warnings (0.01 sec)

mysql> select a and b as c, a, b from tt order by c;
+---+------+------+
| c | a    | b    |
+---+------+------+
| NULL | NULL | NULL |
| NULL |    7 | NULL |
| NULL | NULL | 2    |
| NULL | NULL | 2    |
| 0 | NULL | w    |
| 0 |    0 | 2    |
| 0 |    7 | f    |
| 1 |    7 | 3    |
+---+------+------+
8 rows in set, 2 warnings (0.00 sec)
```

The cause of this problem is that when the order by expression in the `topN` executor is `logicalAnd` or `logicalOr`, `keepNull` should be set to true when the `topN` executor is pushed down.

### Check List 

Tests 

- Unit test
- Integration test

### CODE REVIEW

**Comment by XuHuaiyu** [comment]
@Reminiscent CI failed, please check it.

**Comment by Reminiscent** [comment]
@XuHuaiyu  @breeswish  PTAL. Please review https://github.com/pingcap/tipb/pull/176 first. thx!

File: expression/distsql_builtin.go [hunk 1]
```diff
@@ -41,103 +44,111 @@ func PbTypeToFieldType(tp *tipb.FieldType) *types.FieldType {
...
+func getSignatureByPB(ctx sessionctx.Context, sigCode tipb.ScalarFuncSig, tp *tipb.FieldType, metadata []byte, args []Expression) (f builtinFunc, e error) {
 	fieldTp := PbTypeToFieldType(tp)
 	base := newBaseBuiltinFunc(ctx, args)
 	base.tp = fieldTp
+	var message proto.Message
+	if len(metadata) != 0 {
+		err := proto.Unmarshal(metadata, message)
+		if err != nil {
+			logutil.BgLogger().Error("decode metadata", zap.Any("metadata", metadata), zap.Error(err))
+			return nil, nil
```
Comment (XuHuaiyu) [inline expression/distsql_builtin.go:56]: Should we return the error here?

Comment (Reminiscent) [reply expression/distsql_builtin.go:56]: In the [previous implementation](https://github.com/pingcap/tidb/blob/master/expression/expr_to_pb.go#L252-L259), we just return `nil`. @XuHuaiyu PTAL

**Comment by XuHuaiyu** [comment]
Is this expectant?

``` go
[2020-04-16T06:33:09.000Z] FAIL: const_test.go:180: testMySQLConstSuite.TestHighNotPrecedenceMode

[2020-04-16T06:33:09.000Z] 

[2020-04-16T06:33:09.000Z] const_test.go:191:

[2020-04-16T06:33:09.000Z]     r = tk.MustQuery(`SELECT * FROM t1 WHERE NOT a BETWEEN 2 AND 3;`)

[2020-04-16T06:33:09.000Z] /home/jenkins/agent/workspace/tidb_ghpr_check_2@2/go/src/github.com/pingcap/tidb/util/testkit/testkit.go:288:

[2020-04-16T06:33:09.000Z]     tk.c.Check(err, check.IsNil, comment)

[2020-04-16T06:33:09.000Z] ... value *errors.fundamental = runtime error: invalid memory address or nil pointer dereference ("runtime error: invalid memory address or nil pointer dereference")

[2020-04-16T06:33:09.000Z] ... sql:SELECT * FROM t1 WHERE NOT a BETWEEN 2 AND 3;, args:[]
```

**Comment by XuHuaiyu** [comment]
``` go
[2020-04-16T09:05:14.887Z] PANIC: distsql_builtin_test.go:159: testEvalSuite.TestEval

[2020-04-16T09:05:14.887Z] 

[2020-04-16T09:05:14.887Z] ... Panic: interface conversion: proto.Message is nil, not *tipb.InUnionMetadata (PC=0xFAA711)

[2020-04-16T09:05:14.887Z] 

[2020-04-16T09:05:14.887Z] /usr/local/go/src/runtime/panic.go:679

[2020-04-16T09:05:14.887Z]   in gopanic

[2020-04-16T09:05:14.887Z] /usr/local/go/src/runtime/iface.go:255

[2020-04-16T09:05:14.887Z]   in panicdottypeE

[2020-04-16T09:05:14.887Z] /usr/local/go/src/runtime/iface.go:265

[2020-04-16T09:05:14.887Z]   in panicdottypeI

[2020-04-16T09:05:14.887Z] distsql_builtin.go:74

[2020-04-16T09:05:14.887Z]   in getSignatureByPB

[2020-04-16T09:05:14.887Z] distsql_builtin.go:1056

[2020-04-16T09:05:14.887Z]   in newDistSQLFunctionBySig

[2020-04-16T09:05:14.887Z] distsql_builtin.go:1138

[2020-04-16T09:05:14.887Z]   in PBToExpr

[2020-04-16T09:05:14.887Z] distsql_builtin_test.go:856

[2020-04-16T09:05:14.887Z]   in testEvalSuite.TestEval

[2020-04-16T09:05:14.887Z] /usr/local/go/src/reflect/value.go:321

[2020-04-16T09:05:14.887Z]   in Value.Call

[2020-04-16T09:05:14.887Z] /home/jenkins/agent/workspace/tidb_ghpr_check_2@2/go/pkg/mod/github.com/pingcap/check@v0.0.0-20200212061837-5e12011dc712/check.go:850

[2020-04-16T09:05:14.887Z]   in suiteRunner.forkTest.func1

[2020-04-16T09:05:14.887Z] /home/jenkins/agent/workspace/tidb_ghpr_check_2@2/go/pkg/mod/github.com/pingcap/check@v0.0.0-20200212061837-5e12011dc712/check.go:739

[2020-04-16T09:05:14.887Z]   in suiteRunner.forkCall.func1

[2020-04-16T09:05:14.887Z] /usr/local/go/src/runtime/asm_amd64.s:1357

[2020-04-16T09:05:14.887Z]   in goexit
```

**Comment by Reminiscent** [comment]
@XuHuaiyu @breeswish PTAL. Thanks!

**Comment by breeswish** [comment]
Do we need to fix at TiKV side as well? Otherwise the results will be different so that copr-test will fail.

**Comment by Reminiscent** [comment]
After an offline discussion, we decided to use another method to solve this problem. We will close this PR and reopen a new one.

**Pull request closed without merging**